### ISSUE TITLE
How to correctly set up a custom number of classes in Cascade Mask R-CNN

### ISSUE THREAD

**Issue by AndreaPi** [open-mmlab/mmdetection#6340]
Hi,

I wrote a config file to train Cascade Mask R-CNN with my custom dataset of type `COCODataset`. For the sake of testing, I used `balloon` as a custom COCO dataset (1 class). The config file `test_wrong_bbox_head.py`  is as follows:

```python
# base config
_base_ = ['/mmdetection/configs/cascade_rcnn/cascade_mask_rcnn_r50_fpn_20e_coco.py']

#### dataset-related parameters: edit accordingly when changing dataset
# classes
classes = ('balloon',)
n_classes = len(classes)
model = dict(
    roi_head=dict(
        bbox_head =[
            dict(
                type='Shared2FCBBoxHead',
                num_classes=n_classes),
            dict(
                type='Shared2FCBBoxHead',
                num_classes=n_classes),
            dict(
                type='Shared2FCBBoxHead',
                num_classes=n_classes)
        ],
        mask_head=dict(
            type='FCNMaskHead',
            num_classes=n_classes)))
########################################################################### 
# dataset settings
data = dict(
    samples_per_gpu=4,
    workers_per_gpu=2,
    train=dict(
        type='CocoDataset',
        ann_file=
        '/data/balloon/annotations/train.json',
        img_prefix=
        '/data/balloon/images/',
        classes=classes),
    val=dict(
        type='CocoDataset',
        ann_file=
        '/data/balloon/annotations/validation.json',
        img_prefix=
        '/data/balloon/images/',
        classes=classes),
    test=dict(
        type='CocoDataset',
        ann_file=
        '/data/balloon/annotations/validation.json',
        img_prefix=
        '/data/balloon/images/',
        classes=classes),
)

evaluation = dict(metric=['bbox', 'segm'], save_best='bbox_mAP')
# optimizer `settings`
optimizer = dict(type='SGD', lr=0.005)
lr_config = dict(  
    step=[9, 16]) 
# workflow and workdir settings
runner = dict(type='EpochBasedRunner', max_epochs=20)
checkpoint_config = dict(interval=-1) 
workflow = [('train', 1), ('val', 1)] # this prints both train & val loss for each epoch
work_dir = 'logs/test_on_balloon_dataset'# put the path to your logs folder for this experiment here
## logger settings
log_config = dict(
    hooks=[dict(type='TextLoggerHook'),
           dict(type='TensorboardLoggerHook')])
custom_hooks = [dict(type='NumClassCheckHook')]
```

Traning the model with 
```
python /mmdetection/tools/train.py configs/test_wrong_bbox_head.py
```
seems to work nicely. **However**, inspecting the config file with
```
python /mmdetection/tools/misc/print_config.py configs/test_wrong_bbox_head.py
```
reveals that the configuration of the `bbox_head` was completely lost:

```python
Config:
model = dict(
    type='CascadeRCNN',
    backbone=dict(
        type='ResNet',
        depth=50,
        num_stages=4,
        out_indices=(0, 1, 2, 3),
        frozen_stages=1,
        norm_cfg=dict(type='BN', requires_grad=True),
        norm_eval=True,
        style='pytorch',
        init_cfg=dict(type='Pretrained', checkpoint='torchvision://resnet50')),
    neck=dict(
        type='FPN',
        in_channels=[256, 512, 1024, 2048],
        out_channels=256,
        num_outs=5),
.
.
.
      bbox_head=[
            dict(type='Shared2FCBBoxHead', num_classes=1),
            dict(type='Shared2FCBBoxHead', num_classes=1),
            dict(type='Shared2FCBBoxHead', num_classes=1)
        ],
.
.
.
```

As a comparison, the configuration of the `bbox_head` in the base config file is:

```python
        .
        .
        .
        bbox_head=[
            dict(
                type='Shared2FCBBoxHead',
                in_channels=256,
                fc_out_channels=1024,
                roi_feat_size=7,
                num_classes=80,
                bbox_coder=dict(
                    type='DeltaXYWHBBoxCoder',
                    target_means=[0., 0., 0., 0.],
                    target_stds=[0.1, 0.1, 0.2, 0.2]),
                reg_class_agnostic=True,
                loss_cls=dict(
                    type='CrossEntropyLoss',
                    use_sigmoid=False,
                    loss_weight=1.0),
                loss_bbox=dict(type='SmoothL1Loss', beta=1.0,
                               loss_weight=1.0)),
            dict(
                type='Shared2FCBBoxHead',
                in_channels=256,
                fc_out_channels=1024,
                roi_feat_size=7,
                num_classes=80,
                bbox_coder=dict(
                    type='DeltaXYWHBBoxCoder',
                    target_means=[0., 0., 0., 0.],
                    target_stds=[0.05, 0.05, 0.1, 0.1]),
                reg_class_agnostic=True,
                loss_cls=dict(
                    type='CrossEntropyLoss',
                    use_sigmoid=False,
                    loss_weight=1.0),
                loss_bbox=dict(type='SmoothL1Loss', beta=1.0,
                               loss_weight=1.0)),
            dict(
                type='Shared2FCBBoxHead',
                in_channels=256,
                fc_out_channels=1024,
                roi_feat_size=7,
                num_classes=80,
                bbox_coder=dict(
                    type='DeltaXYWHBBoxCoder',
                    target_means=[0., 0., 0., 0.],
                    target_stds=[0.033, 0.033, 0.067, 0.067]),
                reg_class_agnostic=True,
                loss_cls=dict(
                    type='CrossEntropyLoss',
                    use_sigmoid=False,
                    loss_weight=1.0),
                loss_bbox=dict(type='SmoothL1Loss', beta=1.0, loss_weight=1.0))
        ],
```
How can I modify the number of classes in the config file **without** completely mangling the configuration of the `bbox_head`? For now I'm doing it this way (`test_correct_bbox_head.py` config file):

```python
# base config
_base_ = ['/mmdetection/configs/cascade_rcnn/cascade_mask_rcnn_r50_fpn_20e_coco.py']

#### dataset-related parameters: edit accordingly when changing dataset
# classes
classes = ('balloon',)
n_classes = len(classes)
model = dict(
    roi_head=dict(
        bbox_head=[
            dict(
                type='Shared2FCBBoxHead',
                in_channels=256,
                fc_out_channels=1024,
                roi_feat_size=7,
                num_classes=n_classes,
                bbox_coder=dict(
                    type='DeltaXYWHBBoxCoder',
                    target_means=[0.0, 0.0, 0.0, 0.0],
                    target_stds=[0.1, 0.1, 0.2, 0.2]),
                reg_class_agnostic=True,
                loss_cls=dict(
                    type='CrossEntropyLoss',
                    use_sigmoid=False,
                    loss_weight=1.0),
                loss_bbox=dict(type='SmoothL1Loss', beta=1.0,
                               loss_weight=1.0)),
            dict(
                type='Shared2FCBBoxHead',
                in_channels=256,
                fc_out_channels=1024,
                roi_feat_size=7,
                num_classes=n_classes,
                bbox_coder=dict(
                    type='DeltaXYWHBBoxCoder',
                    target_means=[0.0, 0.0, 0.0, 0.0],
                    target_stds=[0.05, 0.05, 0.1, 0.1]),
                reg_class_agnostic=True,
                loss_cls=dict(
                    type='CrossEntropyLoss',
                    use_sigmoid=False,
                    loss_weight=1.0),
                loss_bbox=dict(type='SmoothL1Loss', beta=1.0,
                               loss_weight=1.0)),
            dict(
                type='Shared2FCBBoxHead',
                in_channels=256,
                fc_out_channels=1024,
                roi_feat_size=7,
                num_classes=n_classes,
                bbox_coder=dict(
                    type='DeltaXYWHBBoxCoder',
                    target_means=[0.0, 0.0, 0.0, 0.0],
                    target_stds=[0.033, 0.033, 0.067, 0.067]),
                reg_class_agnostic=True,
                loss_cls=dict(
                    type='CrossEntropyLoss',
                    use_sigmoid=False,
                    loss_weight=1.0),
                loss_bbox=dict(type='SmoothL1Loss', beta=1.0, loss_weight=1.0))
        ],
        mask_head=dict(
            type='FCNMaskHead',
            num_classes=n_classes)))
########################################################################### 
```

This works, but it's very cumbersome because I basically have to duplicate a large portion of the `_base_` config in my child config. Is there a better way? I attach both config files for your reference in [Archive.zip](https://github.com/open-mmlab/mmdetection/files/7391085/Archive.zip).

**Comment by BIGWangYuDong** [comment]
it is suggested to use `test_correct_bbox_head`
you should set all parameters in the list of cascade head
